### PR TITLE
✨ Source MongoDB V2: Filter included fields during conversion to Airbyte record

### PR DESCRIPTION
**Pull request by jdpgrailsdev**
## What
* The Debezium MongoDB connector only supports a field exclude list, instead of an include list.  This means that collections for which we did not discover all known fields, we may publish records with fields not present in the configured catalog/discovered catalog during the Debezium incremental iterator. 

## How
* This change reverses the behavior to be an explicit check to see if the field in the Debezium event that is about to be converted to an Airbyte record is present in the discovered catalog instead of excluded fields not in the catalog.
* Deletes the existing infrastructure to build the exclude list

## Recommended reading order
1. `DebeziumEventUtils.java`
2. `MongoDbCdcEventUtils.java`
3. `MongoDbCdcInitializer.java`

Unit test cases have been added to verify the new include filtering logic based on the configured catalog.  In addition, all existing unit tests, both in the shared Debezium code and the MongoDB source pass, as well as the integration tests in the MongoDB source.

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-mongodb-v2 test report (commit [`647bd54bd8`](https://github.com/airbytehq/airbyte/pull/30942/commits/647bd54bd8b21da85c9b7843391ddae3b12f2ec7)) - ❌

⏲️  Total pipeline duration: 07mn08s 

| Step                                                           | Result   |
|:---------------------------------------------------------------|:---------|
| Build connector tar                                            | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64 | ✅        |
| Java Connector Unit Tests                                      | ✅        |
| Java Connector Integration Tests                               | ❌        |
| Acceptance tests                                               | ❌        |
| Validate metadata for source-mongodb-v2                        | ✅        |
| Connector version semver check                                 | ✅        |
| Connector version increment check                              | ❌        |
| QA checks                                                      | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/jonathan_mongodb-debezium-include-fields/1696020163/647bd54bd8b21da85c9b7843391ddae3b12f2ec7/source-mongodb-v2/0.2.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:647bd54bd8b21da85c9b7843391ddae3b12f2ec7)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mongodb-v2 test report (commit [`245698c7a4`](https://github.com/airbytehq/airbyte/pull/30942/commits/245698c7a45aa381926e7c2a70b4dc70af3d5729)) - ❌

⏲️  Total pipeline duration: 08mn17s 

| Step                                                           | Result   |
|:---------------------------------------------------------------|:---------|
| Build connector tar                                            | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64 | ✅        |
| Java Connector Unit Tests                                      | ✅        |
| Java Connector Integration Tests                               | ❌        |
| Acceptance tests                                               | ❌        |
| Validate metadata for source-mongodb-v2                        | ✅        |
| Connector version semver check                                 | ✅        |
| Connector version increment check                              | ❌        |
| QA checks                                                      | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/jonathan_mongodb-debezium-include-fields/1696259957/245698c7a45aa381926e7c2a70b4dc70af3d5729/source-mongodb-v2/0.2.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:245698c7a45aa381926e7c2a70b4dc70af3d5729)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumEventUtils.java [hunk 1]
```diff
@@ -58,22 +62,24 @@ private static AirbyteMessage buildAirbyteMessage(final JsonNode source,
...
                                                    final CdcMetadataInjector cdcMetadataInjector,
+                                                   final ConfiguredAirbyteCatalog configuredAirbyteCatalog,
                                                    final Instant emittedAt) {
     final JsonNode debeziumEventKey = event.eventKeyAsJson();
     final JsonNode debeziumEvent = event.eventValueAsJson();
     final JsonNode before = debeziumEvent.get(BEFORE_EVENT);
     final JsonNode after = debeziumEvent.get(AFTER_EVENT);
     final JsonNode source = debeziumEvent.get(SOURCE_EVENT);
     final String operation = debeziumEvent.get(OPERATION_FIELD).asText();
+    final Set<String> configuredFields = getConfiguredMongoDbCollectionFields(source, configuredAirbyteCatalog, cdcMetadataInjector);
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumEventUtils.java:73]: @jpefaur I don't think so.  @subodh1810 and I chatted about this during the sync meeting.  We are already doing most of this work anyway to convert the datatype.  Adding the additional lines of code to check if the field that we are converting is included should not make a difference.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mongodb/MongoDbCdcEventUtils.java [hunk 2]
```diff
@@ -133,27 +134,34 @@ private static void formatDocument(final Document document, final ObjectNode obj
...
     reader.readStartDocument();
     while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
       final var fieldName = reader.readName();
       final var fieldType = reader.getCurrentBsonType();
-      if (DOCUMENT.equals(fieldType)) {
-        // recursion in used to parse inner documents
-        jsonNodes.set(fieldName, readDocument(reader, (ObjectNode) Jsons.jsonNode(Collections.emptyMap()), columnNames));
-      } else if (ARRAY.equals(fieldType)) {
-        jsonNodes.set(fieldName, readArray(reader, columnNames, fieldName));
+      if (columnNames.contains(fieldName) || columnNames.contains(ALLOW_ALL)) {
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mongodb/MongoDbCdcEventUtils.java:142]: @jpefaur The code goes recursively through all of the data to convert the data type.  However, we only check the top level fields for inclusion (and for discovery).  Therefore, to ensure that any nested objects are still included/converted, I needed a mechanism to let those fields pass the inclusion filter.  That is why it is using allow all for those situations.  I can add a better clarifying comment.

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumEventUtils.java [hunk 3]
```diff
@@ -30,10 +33,11 @@ public class DebeziumEventUtils {
 
   public static AirbyteMessage toAirbyteMessage(final ChangeEventWithMetadata event,
                                                 final CdcMetadataInjector cdcMetadataInjector,
+                                                final ConfiguredAirbyteCatalog configuredAirbyteCatalog,
```
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumEventUtils.java:36]: @jpefaur Agreed.  This whole thing probably needs to be refactored into a parent class and subclasses given that MongoDB is different enough from the relational database flows.  This should probably be something that is done in short order after getting this change in.  cc: @rodireich

**Comment by airbyte-oss-build-runner** [comment]
##  source-mongodb-v2 test report (commit [`00409d455e`](https://github.com/airbytehq/airbyte/pull/30942/commits/00409d455e249c34284bda13b46ee2b7d27c18a9)) - ❌

⏲️  Total pipeline duration: 06mn45s 

| Step                                                           | Result   |
|:---------------------------------------------------------------|:---------|
| Build connector tar                                            | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64 | ✅        |
| Java Connector Unit Tests                                      | ✅        |
| Java Connector Integration Tests                               | ❌        |
| Acceptance tests                                               | ❌        |
| Validate metadata for source-mongodb-v2                        | ✅        |
| Connector version semver check                                 | ✅        |
| Connector version increment check                              | ❌        |
| QA checks                                                      | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/jonathan_mongodb-debezium-include-fields/1696264295/00409d455e249c34284bda13b46ee2b7d27c18a9/source-mongodb-v2/0.2.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:00409d455e249c34284bda13b46ee2b7d27c18a9)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-mongodb-v2 test report (commit [`7425c9f21c`](https://github.com/airbytehq/airbyte/pull/30942/commits/7425c9f21c46942596b45fc586e5d1e92a830f5f)) - ❌

⏲️  Total pipeline duration: 11mn12s 

| Step                                                           | Result   |
|:---------------------------------------------------------------|:---------|
| Build connector tar                                            | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64 | ✅        |
| Java Connector Unit Tests                                      | ✅        |
| Java Connector Integration Tests                               | ❌        |
| Acceptance tests                                               | ❌        |
| Validate metadata for source-mongodb-v2                        | ✅        |
| Connector version semver check                                 | ✅        |
| Connector version increment check                              | ❌        |
| QA checks                                                      | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/jonathan_mongodb-debezium-include-fields/1696269108/7425c9f21c46942596b45fc586e5d1e92a830f5f/source-mongodb-v2/0.2.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:7425c9f21c46942596b45fc586e5d1e92a830f5f)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

**Review by jpefaur — COMMENTED** [review]
one more comment!

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mongodb/MongoDbCdcEventUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mongodb/MongoDbCdcEventUtils.java:149]: Yeah...I could extract the check to a method and pass in an override boolean value.  Will fix.

**Review by jpefaur — APPROVED** [review]

**Comment by airbyte-oss-build-runner** [comment]
##  source-mongodb-v2 test report (commit [`aa680d98ec`](https://github.com/airbytehq/airbyte/pull/30942/commits/aa680d98ec1055471643b34903a692fb22d2bfd1)) - ❌

⏲️  Total pipeline duration: 12mn55s 

| Step                                                           | Result   |
|:---------------------------------------------------------------|:---------|
| Build connector tar                                            | ✅        |
| Build source-mongodb-v2 docker image for platform linux/x86_64 | ✅        |
| Java Connector Unit Tests                                      | ✅        |
| Java Connector Integration Tests                               | ❌        |
| Acceptance tests                                               | ❌        |
| Validate metadata for source-mongodb-v2                        | ✅        |
| Connector version semver check                                 | ✅        |
| Connector version increment check                              | ❌        |
| QA checks                                                      | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/jonathan_mongodb-debezium-include-fields/1696282593/aa680d98ec1055471643b34903a692fb22d2bfd1/source-mongodb-v2/0.2.5/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:aa680d98ec1055471643b34903a692fb22d2bfd1)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mongodb-v2 test
```

**Pull request merged**